File: endpoint.py
#!/usr/bin/env python3
"""
ZeroTrust DNS Endpoint (Python version)
Provides local DNS proxy with mTLS authentication to centralized DNS server
"""

import json
import logging
import socket
import ssl
import struct
import sys
import threading
from pathlib import Path

from jwcrypto import jwk, jwt
from cryptography import x509
from cryptography.hazmat.backends import default_backend

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
log = logging.getLogger(__name__)


class ZeroTrustEndpoint:
    def __init__(self):
        self.config = None
        self.tls_context = None
        self.local_port = 53

    def load_config(self):
        """Load and validate JWT-signed configuration"""
        try:
            # Read JWT token
            zt_token = Path("config.zt").read_text()

            # Read CA certificate for JWT verification
            ca_pem = Path("ca.crt").read_bytes()
            ca_cert = x509.load_pem_x509_certificate(ca_pem, default_backend())

            # Extract RSA public key from CA cert
            ca_key = jwk.JWK.from_pyca(ca_cert.public_key())

            # Verify and decode JWT
            token = jwt.JWT(key=ca_key, jwt=zt_token)
            claims = json.loads(token.claims)

            # Parse config payload
            self.config = json.loads(claims["data"])
            log.info(
                f"Loaded config: Type={self.config['type']}, Server={self.config['server']}"
            )

        except FileNotFoundError as e:
            log.fatal(f"Missing required file: {e.filename}")
            sys.exit(1)
        except Exception as e:
            log.fatal(f"Failed to load config: {e}")
            sys.exit(1)

    def setup_tls(self):
        """Configure mTLS context with client certificate"""
        try:
            # Create SSL context for client authentication
            self.tls_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)

            # Load client certificate and key
            self.tls_context.load_cert_chain(
                certfile="endpoint.crt", keyfile="endpoint.key"
            )

            # Load CA certificate for server verification
            self.tls_context.load_verify_locations("ca.crt")

            # Set server hostname for SNI
            self.tls_context.check_hostname = True

            # Require TLS 1.3
            self.tls_context.minimum_version = ssl.TLSVersion.TLSv1_3

            log.info("mTLS context initialized")

        except ssl.SSLError as e:
            # Check if endpoint.crt is actually a CSR
            try:
                crt_data = Path("endpoint.crt").read_text()
                if "CERTIFICATE REQUEST" in crt_data:
                    log.fatal(
                        "endpoint.crt is a CSR, not a signed certificate. Regenerate client after server update."
                    )
                    sys.exit(1)
            except:
                pass
            log.fatal(f"Failed to setup TLS: {e}")
            sys.exit(1)
        except Exception as e:
            log.fatal(f"Failed to load certificates: {e}")
            sys.exit(1)

    def start_local_dns(self):
        """Start local UDP DNS server on 127.0.0.1"""
        # Try port 53 first (requires root/CAP_NET_BIND_SERVICE)
        for port in [53, 5353]:
            try:
                sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                sock.bind(("127.0.0.1", port))
                self.local_port = port

                if port == 5353:
                    log.warning(
                        f"Permission denied on port 53, using port {port} (run as root or grant CAP_NET_BIND_SERVICE for port 53)"
                    )

                log.info(f"Local DNS listening on 127.0.0.1:{port}")

                # Handle incoming DNS queries
                while True:
                    try:
                        data, client_addr = sock.recvfrom(65536)
                        # Handle each query in separate thread
                        threading.Thread(
                            target=self.handle_query,
                            args=(data, client_addr, sock),
                            daemon=True,
                        ).start()
                    except Exception as e:
                        log.error(f"Error receiving query: {e}")

            except PermissionError:
                if port == 53:
                    continue  # Try 5353
                else:
                    log.fatal(f"Failed to bind to port {port}: Permission denied")
                    sys.exit(1)
            except Exception as e:
                log.fatal(f"Failed to start DNS server: {e}")
                sys.exit(1)

    def handle_query(self, query: bytes, client_addr: tuple, sock: socket.socket):
        """Process DNS query and return response"""
        try:
            # Service endpoints first try public DNS for non-private queries
            if self.config.get("type") == "service":
                public_response = self.try_public_dns(query)
                if public_response:
                    sock.sendto(public_response, client_addr)
                    return

            # Forward to ZeroTrust DNS server via mTLS
            response = self.forward_to_server(query)
            if response:
                sock.sendto(response, client_addr)

        except Exception as e:
            log.error(f"Error handling query: {e}")

    def try_public_dns(self, query: bytes) -> bytes:
        """Try resolving via public DNS (Cloudflare) for service endpoints"""
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.settimeout(2.0)
            sock.sendto(query, ("1.1.1.1", 53))
            response, _ = sock.recvfrom(512)
            sock.close()

            # Verify we got a valid response (more than just header)
            if len(response) > 12:
                return response

        except socket.timeout:
            pass
        except Exception as e:
            log.debug(f"Public DNS query failed: {e}")

        return None

    def forward_to_server(self, query: bytes) -> bytes:
        """Forward DNS query to ZeroTrust server via mTLS DoT (DNS over TLS)"""
        try:
            # Parse server address
            server_host, server_port = self.config["server"].rsplit(":", 1)
            server_port = int(server_port)

            # Create TLS connection with mTLS authentication
            with socket.create_connection(
                (server_host, server_port), timeout=5.0
            ) as raw_sock:
                with self.tls_context.wrap_socket(
                    raw_sock, server_hostname=self.config["server_name"]
                ) as tls_sock:
                    # Send DNS query with 2-byte length prefix (RFC 7858 - DNS over TLS)
                    length_prefix = struct.pack("!H", len(query))
                    tls_sock.sendall(length_prefix + query)

                    # Read response length
                    resp_len_bytes = self._recv_exact(tls_sock, 2)
                    if not resp_len_bytes:
                        return None
                    resp_len = struct.unpack("!H", resp_len_bytes)[0]

                    # Read response data
                    response = self._recv_exact(tls_sock, resp_len)
                    return response

        except ssl.SSLError as e:
            log.error(f"TLS error connecting to server: {e}")
        except socket.timeout:
            log.error("Connection to server timed out")
        except Exception as e:
            log.error(f"Failed to forward query: {e}")

        return None

    def _recv_exact(self, sock: socket.socket, n: int) -> bytes:
        """Read exactly n bytes from socket"""
        data = b""
        while len(data) < n:
            chunk = sock.recv(n - len(data))
            if not chunk:
                return None
            data += chunk
        return data

    def run(self):
        """Main entry point"""
        log.info("ZeroTrust DNS Endpoint starting...")

        # Load configuration
        self.load_config()

        # Setup mTLS
        self.setup_tls()

        # Log endpoint type and target server
        endpoint_type = self.config.get("type", "unknown").upper()
        log.info(f"ZeroTrust {endpoint_type} Endpoint Active → {self.config['server']}")

        if self.config.get("proxy"):
            log.info(f"Service proxy available at: {self.config['proxy']}")

        if self.config.get("domains"):
            log.info(f"Authorized domains: {', '.join(self.config['domains'])}")

        # Start local DNS server (blocking)
        self.start_local_dns()


def main():
    """Entry point"""
    # Check required files exist
    required_files = ["config.zt", "ca.crt", "endpoint.crt", "endpoint.key"]
    missing = [f for f in required_files if not Path(f).exists()]

    if missing:
        log.fatal(f"Missing required files: {', '.join(missing)}")
        log.fatal("Ensure you've extracted all files from the downloaded ZIP package")
        sys.exit(1)

    # Start endpoint
    endpoint = ZeroTrustEndpoint()

    try:
        endpoint.run()
    except KeyboardInterrupt:
        log.info("\nShutting down...")
        sys.exit(0)
    except Exception as e:
        log.fatal(f"Fatal error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: endpoint.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"os"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type Config struct {
	Server     string   `json:"server"`
	Proxy      string   `json:"proxy"`
	ServerName string   `json:"server_name"`
	Type       string   `json:"type"`
	Domains    []string `json:"domains"`
	Expires    string   `json:"expires"`
}

type JWTClaims struct {
	Data string `json:"data"`
	jwt.RegisteredClaims
}

func loadConfig() (*Config, error) {
	// Read JWT token
	ztToken, err := os.ReadFile("config.zt")
	if err != nil {
		return nil, fmt.Errorf("failed to read config.zt: %v", err)
	}

	// Read CA certificate for verification
	caPEM, err := os.ReadFile("ca.crt")
	if err != nil {
		return nil, fmt.Errorf("failed to read ca.crt: %v", err)
	}

	// Parse CA certificate
	block, _ := x509.ParseCertificate(caPEM)
	if block == nil {
		// Try PEM format
		certPool := x509.NewCertPool()
		if !certPool.AppendCertsFromPEM(caPEM) {
			return nil, fmt.Errorf("failed to parse CA certificate")
		}
		// Get first cert from pool for public key
		caCert, err := x509.ParseCertificate(caPEM)
		if err != nil {
			// Parse PEM block
			block, _ := parsePEMCertificate(caPEM)
			caCert = block
		}
		block = caCert
	}

	// Parse and verify JWT
	token, err := jwt.ParseWithClaims(string(ztToken), &JWTClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return block.PublicKey, nil
	})

	if err != nil {
		return nil, fmt.Errorf("failed to parse JWT: %v", err)
	}

	claims, ok := token.Claims.(*JWTClaims)
	if !ok || !token.Valid {
		return nil, fmt.Errorf("invalid token")
	}

	// Parse config from JWT data
	var config Config
	if err := json.Unmarshal([]byte(claims.Data), &config); err != nil {
		return nil, fmt.Errorf("failed to parse config: %v", err)
	}

	return &config, nil
}

func parsePEMCertificate(pemData []byte) (*x509.Certificate, error) {
	// Simple PEM parser
	start := []byte("-----BEGIN CERTIFICATE-----")
	end := []byte("-----END CERTIFICATE-----")
	
	startIdx := findIndex(pemData, start)
	endIdx := findIndex(pemData, end)
	
	if startIdx == -1 || endIdx == -1 {
		return nil, fmt.Errorf("invalid PEM format")
	}
	
	// Extract base64 data between markers
	b64Data := pemData[startIdx+len(start):endIdx]
	
	// Decode base64 (simplified - you'd use encoding/base64 in production)
	derData := decodeBase64(b64Data)
	
	return x509.ParseCertificate(derData)
}

func findIndex(data []byte, pattern []byte) int {
	for i := 0; i <= len(data)-len(pattern); i++ {
		match := true
		for j := 0; j < len(pattern); j++ {
			if data[i+j] != pattern[j] {
				match = false
				break
			}
		}
		if match {
			return i
		}
	}
	return -1
}

func decodeBase64(data []byte) []byte {
	// Simplified base64 decoder - in production use encoding/base64
	// This is just a placeholder
	return data
}

func setupTLS(config *Config) (*tls.Config, error) {
	// Load client certificate
	cert, err := tls.LoadX509KeyPair("endpoint.crt", "endpoint.key")
	if err != nil {
		return nil, fmt.Errorf("failed to load client certificate: %v", err)
	}

	// Load CA certificate
	caCert, err := os.ReadFile("ca.crt")
	if err != nil {
		return nil, fmt.Errorf("failed to read CA certificate: %v", err)
	}

	caCertPool := x509.NewCertPool()
	if !caCertPool.AppendCertsFromPEM(caCert) {
		return nil, fmt.Errorf("failed to parse CA certificate")
	}

	tlsConfig := &tls.Config{
		Certificates: []tls.Certificate{cert},
		RootCAs:      caCertPool,
		ServerName:   config.ServerName,
		MinVersion:   tls.VersionTLS13,
	}

	return tlsConfig, nil
}

func startLocalDNS(config *Config, tlsConfig *tls.Config) {
	// Try port 53 first (requires root/admin)
	ports := []int{53, 5353}
	var conn *net.UDPConn
	var listenPort int

	for _, port := range ports {
		addr := &net.UDPAddr{
			IP:   net.ParseIP("127.0.0.1"),
			Port: port,
		}
		var err error
		conn, err = net.ListenUDP("udp", addr)
		if err == nil {
			listenPort = port
			if port == 5353 {
				log.Printf("Warning: Could not bind to port 53, using port %d (run as root/admin for port 53)", port)
			}
			break
		}
		if port == ports[len(ports)-1] {
			log.Fatalf("Failed to bind to any DNS port: %v", err)
		}
	}
	defer conn.Close()

	log.Printf("Local DNS listening on 127.0.0.1:%d", listenPort)

	buffer := make([]byte, 512)
	for {
		n, clientAddr, err := conn.ReadFromUDP(buffer)
		if err != nil {
			log.Printf("Error reading from UDP: %v", err)
			continue
		}

		go handleDNSQuery(conn, clientAddr, buffer[:n], config, tlsConfig)
	}
}

func handleDNSQuery(conn *net.UDPConn, clientAddr *net.UDPAddr, query []byte, config *Config, tlsConfig *tls.Config) {
	// For service endpoints, try public DNS first
	if config.Type == "service" {
		response := tryPublicDNS(query)
		if response != nil {
			conn.WriteToUDP(response, clientAddr)
			return
		}
	}

	// Forward to ZeroTrust DNS server via mTLS
	response := forwardToServer(query, config, tlsConfig)
	if response != nil {
		conn.WriteToUDP(response, clientAddr)
	}
}

func tryPublicDNS(query []byte) []byte {
	conn, err := net.DialTimeout("udp", "1.1.1.1:53", 2*time.Second)
	if err != nil {
		return nil
	}
	defer conn.Close()

	conn.SetDeadline(time.Now().Add(2 * time.Second))
	
	if _, err := conn.Write(query); err != nil {
		return nil
	}

	buffer := make([]byte, 512)
	n, err := conn.Read(buffer)
	if err != nil {
		return nil
	}

	if n > 12 { // Valid DNS response
		return buffer[:n]
	}

	return nil
}

func forwardToServer(query []byte, config *Config, tlsConfig *tls.Config) []byte {
	// Connect to DNS server with mTLS
	dialer := &net.Dialer{
		Timeout: 5 * time.Second,
	}

	conn, err := tls.DialWithDialer(dialer, "tcp", config.Server, tlsConfig)
	if err != nil {
		log.Printf("Failed to connect to DNS server: %v", err)
		return nil
	}
	defer conn.Close()

	// Send DNS query with 2-byte length prefix (RFC 7858 - DNS over TLS)
	length := uint16(len(query))
	lengthBytes := []byte{byte(length >> 8), byte(length & 0xff)}
	
	if _, err := conn.Write(append(lengthBytes, query...)); err != nil {
		log.Printf("Failed to send DNS query: %v", err)
		return nil
	}

	// Read length-prefixed DNS response
	respLenBuf := make([]byte, 2)
	if _, err := conn.Read(respLenBuf); err != nil {
		log.Printf("Failed to read DNS response length: %v", err)
		return nil
	}

	respLen := int(respLenBuf[0])<<8 | int(respLenBuf[1])
	if respLen <= 0 || respLen > 4096 {
		log.Printf("Invalid DNS response length: %d", respLen)
		return nil
	}

	resp := make([]byte, respLen)
	if _, err := conn.Read(resp); err != nil {
		log.Printf("Failed to read DNS response: %v", err)
		return nil
	}

	return resp
}

func main() {
	config, err := loadConfig()
	if err != nil {
		log.Fatalf("Failed to load config: %v", err)
	}

	tlsConfig, err := setupTLS(config)
	if err != nil {
		log.Fatalf("Failed to set up TLS: %v", err)
	}

	startLocalDNS(config, tlsConfig)
}